torch/csrc/autograd/python_variable.cpp

```diff
@@ -0,0 +1,386 @@
+  Py_INCREF(data);
+  Py_XDECREF(self->data);
+  self->data = data;
+  auto& var = *self->cdata;
+  auto tensor = torch::createTensor(data);
+  var.data.swap(tensor);
+  return 0;
+}
+
+PyObject *THPVariable_get_raw_grad(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  if (!var.grad) {
+    Py_RETURN_NONE;
+  }
+  return THPVariable_Wrap(var.grad);
+}
+
+int THPVariable_set_raw_grad(THPVariable *self, PyObject *data)
+{
+  auto& var = *self->cdata;
+  if (data == Py_None) {
+    var.grad.reset();
+    return 0;
+  }
+  THPUtils_assertRet(-1, THPVariable_Check(data),
+      "expected Variable or None (got %s)", THPUtils_typename(data));
+  var.grad = ((THPVariable*)data)->cdata;
+  return 0;
+}
+
+PyObject *THPVariable_get_grad(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  if (!var.grad) {
+#ifdef WITH_CUDA
+    THCPAutoGPU __guard(var.data->getDevice());
+#endif
+    auto grad = var.data->newTensor();
+    grad->resizeAs(*var.data).zero();
+    var.grad = std::make_shared<Variable>(std::move(grad), 0, 1);
+  }
+  return THPVariable_Wrap(var.grad);
+}
+
+PyObject *THPVariable_get_volatile(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  return PyBool_FromLong(var.is_volatile);
+}
+
+int THPVariable_set_volatile(THPVariable *self, PyObject *obj)
+{
+  THPUtils_assertRet(-1, PyBool_Check(obj), "volatile must be a bool");
+  THPUtils_assertRet(-1, !self->cdata->creator,
+      "volatile can only be set on leaf variables");
+  auto& var = *self->cdata;
+  var.is_volatile = (obj == Py_True);
+  return 0;
+}
+
+PyObject *THPVariable_get_output_nr(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  return PyInt_FromLong(var.output_nr);
+}
+
+PyObject *THPVariable_get_requires_grad(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  return PyBool_FromLong(var.requires_grad);
+}
+
+int THPVariable_set_requires_grad(THPVariable *self, PyObject *obj)
+{
+  THPUtils_assertRet(-1, PyBool_Check(obj), "requires_grad must be a bool");
+  auto& var = *self->cdata;
+  if (var.creator) {
+    const char *hint = "";
+    if (obj == Py_False) {
+      hint = " If you want to use a computed variable in a subgraph "
+             "that doesn't require differentiation use "
+             "var_no_grad = var.detach().";
+    }
+    THPUtils_setError("you can only change requires_grad flags of leaf variables.%s", hint);
+    return -1;
+  }
+  var.requires_grad = (obj == Py_True);
+  return 0;
+}
+
+struct PyVariableHook : public VariableHook {
+  PyVariableHook(PyObject* dict) : dict(dict) {
+    Py_INCREF(dict);
+  }
+  ~PyVariableHook() {
+    AutoGIL gil;
+    Py_DECREF(dict);
+  }
+
+  std::unique_ptr<thpp::Tensor> operator()(const thpp::Tensor& tensor) override {
+    AutoGIL gil;
+    THPObjectPtr grad = torch::createPyObject(tensor);
+    PyObject *key, *value;
+    Py_ssize_t pos = 0;
+    while (PyDict_Next(dict, &pos, &key, &value)) {
+      THPObjectPtr res = PyObject_CallFunctionObjArgs(value, grad.get(), nullptr);
```

Comment:
You're right. This is actually a bug in the current implementation too. (Leaf variable hooks are called with tensors instead of variables)